// server/calculate.js
export const ATOMIC_WEIGHTS = {
  H: 1.008,
  He: 4.0026,
  Li: 6.94,
  Be: 9.0122,
  B: 10.81,
  C: 12.011,
  N: 14.007,
  O: 15.999,
  F: 18.998,
  Ne: 20.18,
  Na: 22.99,
  Mg: 24.305,
  Al: 26.982,
  Si: 28.085,
  P: 30.974,
  S: 32.06,
  Cl: 35.45,
  Ar: 39.948,
  K: 39.098,
  Ca: 40.078,
  Sc: 44.956,
  Ti: 47.867,
  V: 50.942,
  Cr: 51.996,
  Mn: 54.938,
  Fe: 55.845,
  Co: 58.933,
  Ni: 58.693,
  Cu: 63.546,
  Zn: 65.38,
  Ga: 69.723,
  Ge: 72.63,
  As: 74.922,
  Se: 78.971,
  Br: 79.904,
  Kr: 83.798,
  Rb: 85.468,
  Sr: 87.62,
  Y: 88.906,
  Zr: 91.224,
  Nb: 92.906,
  Mo: 95.95,
  Tc: 98,
  Ru: 101.07,
  Rh: 102.91,
  Pd: 106.42,
  Ag: 107.87,
  Cd: 112.41,
  In: 114.82,
  Sn: 118.71,
  Sb: 121.76,
  Te: 127.6,
  I: 126.9,
  Xe: 131.29,
  Cs: 132.91,
  Ba: 137.33,
  La: 138.91,
  Ce: 140.12,
  Pr: 140.91,
  Nd: 144.24,
  Pm: 145,
  Sm: 150.36,
  Eu: 151.96,
  Gd: 157.25,
  Tb: 158.93,
  Dy: 162.5,
  Ho: 164.93,
  Er: 167.26,
  Tm: 168.93,
  Yb: 173.05,
  Lu: 174.97,
  Hf: 178.49,
  Ta: 180.95,
  W: 183.84,
  Re: 186.21,
  Os: 190.23,
  Ir: 192.22,
  Pt: 195.08,
  Au: 196.97,
  Hg: 200.59,
  Tl: 204.38,
  Pb: 207.2,
  Bi: 208.98,
  Po: 209,
  At: 210,
  Rn: 222,
  Fr: 223,
  Ra: 226,
  Ac: 227,
  Th: 232.04,
  Pa: 231.04,
  U: 238.03,
  Np: 237,
  Pu: 244,
  Am: 243,
  Cm: 247,
  Bk: 247,
  Cf: 251,
};

export const IONIC_RADII = {
  H: { "-1": 139.9 },
  Li: { "1": 76 },
  Be: { "2": 45 },
  B: { "3": 27 },
  C: { "4": 16 },
  N: { "3": 16, "5": 13 },
  Na: { "1": 102 },
  Mg: { "2": 72 },
  Al: { "3": 53.5 },
  Si: { "4": 40 },
  P: { "3": 44, "5": 38 },
  S: { "4": 37, "6": 29 },
  Cl: { "5": 12, "7": 27 },
  K: { "1": 138 },
  Ca: { "2": 100 },
  Sc: { "3": 74.5 },
  Ti: { "2": 86, "3": 67, "4": 60.5 },
  V: { "2": 79, "3": 64, "4": 58, "5": 54 },
  Cr: { "2": 73, "3": 61.5, "4": 55, "5": 49, "6": 44 },
  Mn: { "2": 67, "3": 58, "4": 53, "5": 33, "6": 25.5, "7": 46 },
  Fe: { "2": 61, "3": 55, "4": 58.5, "6": 25 },
  Co: { "2": 65, "3": 54.5, "4": 53 },
  Ni: { "2": 69, "3": 56, "4": 48 },
  Cu: { "1": 77, "2": 73, "3": 54 },
  Zn: { "2": 74 },
  Ga: { "3": 62 },
  Ge: { "2": 73, "4": 53 },
  As: { "3": 58, "5": 46 },
  Se: { "4": 50, "6": 42 },
  Br: { "3": 59, "5": 31, "7": 39 },
  Rb: { "1": 152 },
  Sr: { "2": 118 },
  Y: { "3": 90 },
  Zr: { "4": 72 },
  Nb: { "3": 72, "4": 68, "5": 64 },
  Mo: { "3": 69, "4": 65, "5": 61, "6": 59 },
  Tc: { "4": 64.5, "5": 60, "7": 56 },
  Ru: { "3": 68, "4": 62, "5": 56.5, "7": 38, "8": 36 },
  Rh: { "3": 66.5, "4": 60, "5": 55 },
  Pd: { "2": 86, "3": 76, "4": 61.5 },
  Ag: { "1": 115, "2": 94, "3": 75 },
  Cd: { "2": 95 },
  In: { "3": 80 },
  Sn: { "2": 102, "4": 69 },
  Sb: { "3": 76, "5": 60 },
  Te: { "4": 97, "6": 56 },
  I: { "3": 95, "5": 53 },
  Xe: { "8": 48 },
  Cs: { "1": 167 },
  Ba: { "2": 135 },
  La: { "3": 103.2 },
  Ce: { "3": 101, "4": 87 },
  Pr: { "3": 99, "4": 85 },
  Nd: { "3": 98.3 },
  Pm: { "3": 97 },
  Sm: { "3": 95.8 },
  Eu: { "2": 117, "3": 94.7 },
  Gd: { "3": 93.5 },
  Tb: { "3": 92.3, "4": 76 },
  Dy: { "2": 107, "3": 91.2 },
  Ho: { "3": 90.1 },
  Er: { "3": 89 },
  Tm: { "2": 103, "3": 88 },
  Yb: { "2": 102, "3": 86.8 },
  Lu: { "3": 86.1 },
  Hf: { "4": 71 },
  Ta: { "3": 72, "4": 68, "5": 64 },
  W: { "4": 66, "5": 62, "6": 60 },
  Re: { "4": 63, "5": 58, "6": 55, "7": 53 },
  Os: { "4": 63, "5": 57.5, "6": 54.5, "7": 52.5, "8": 39 },
  Ir: { "3": 68, "4": 62.5, "5": 57 },
  Pt: { "2": 80, "4": 62.5, "6": 57 },
  Au: { "1": 137, "3": 85, "5": 57 },
  Hg: { "1": 119, "2": 102 },
  Tl: { "1": 150, "3": 88.5 },
  Pb: { "2": 119, "4": 77.5 },
  Bi: { "3": 103, "5": 76 },
  Po: { "4": 94, "6": 67 },
  Fr: { "1": 180 },
  Ra: { "2": 148 },
  Ac: { "3": 106.5 },
  Th: { "4": 94 },
  Pa: { "3": 104, "4": 90, "5": 78 },
  U: { "3": 102.5, "4": 89, "5": 76, "6": 73 },
  Np: { "3": 110, "4": 101, "5": 87, "6": 75, "7": 72 },
  Pu: { "3": 100, "4": 86, "5": 74, "6": 71 },
  Am: { "2": 126, "3": 97.5, "4": 85 },
  Cm: { "3": 97, "4": 85 },
  Bk: { "3": 96, "4": 83 },
  Cf: { "3": 95, "4": 82.1 },
};

export const NA = 6.02214076e23;
export const RO_CM = 140e-10;

export function parseFormula(formula) {
  const regex = /([A-Z][a-z]?)(\d*)/g;
  const elements = [];
  let match;
  while ((match = regex.exec(formula)) !== null) {
    if (match[1]) {
      elements.push({
        element: match[1],
        count: match[2] ? parseInt(match[2], 10) : 1,
      });
    }
  }
  return elements;
}

export function getMolecularWeight(formula) {
  const elements = parseFormula(formula);
  return elements.reduce((sum, { element, count }) => {
    const w = ATOMIC_WEIGHTS[element];
    if (w === undefined) {
      throw new Error(`Unknown element: ${element}`);
    }
    return sum + w * count;
  }, 0);
}

export function getCationAndOxygen(formula) {
  const elements = parseFormula(formula);
  const oxygenEntry = elements.find((e) => e.element === "O");
  if (!oxygenEntry) {
    throw new Error("FORMULA_NO_OXYGEN");
  }
  const cationEntry = elements.find((e) => e.element !== "O");
  if (!cationEntry) {
    throw new Error("FORMULA_NO_CATION");
  }
  return {
    cation: cationEntry.element,
    n: cationEntry.count,
    m: oxygenEntry.count,
  };
}

export function resolveOxidationState(cation, n, m) {
  const ox = (m * 2) / n;
  return Math.round(ox);
}

export function getCationRadius(cation, oxidation) {
  const radii = IONIC_RADII[cation];
  if (!radii) {
    throw new Error(`No ionic radius data for: ${cation}`);
  }
  const r = radii[String(oxidation)];
  if (r !== undefined) {
    return r;
  }
  const available = Object.keys(radii).map(Number);
  const closest = available.reduce((a, b) =>
    Math.abs(b - oxidation) < Math.abs(a - oxidation) ? b : a
  );
  return radii[String(closest)];
}

export function calculate(compounds) {
  const totalCoeff = compounds.reduce((s, c) => s + c.coefficient, 0);

  const processed = compounds.map((c) => {
    const MW = getMolecularWeight(c.formula);
    const xi = c.coefficient / totalCoeff;
    const { cation, n, m } = getCationAndOxygen(c.formula);
    const oxidation = resolveOxidationState(cation, n, m);
    const R_pm = getCationRadius(cation, oxidation);
    const R_cm = R_pm * 1e-10;
    const Ui_kJ = c.bde_kcal * 4.184;

    const Vi =
      (4 * Math.PI) / 3 * NA * (n * R_cm ** 3 + m * RO_CM ** 3);

    const Gi = (c.density / MW) * Ui_kJ;

    return {
      formula: c.formula,
      MW,
      xi,
      cation,
      oxidation,
      n,
      m,
      R_pm,
      Vi,
      Ui_kJ,
      Gi,
      density: c.density,
    };
  });

  const Vm = processed.reduce((s, c) => s + (c.xi * c.MW) / c.density, 0);

  const sumXiVi = processed.reduce((s, c) => s + c.xi * c.Vi, 0);
  const Vt = sumXiVi / Vm;

  const Gt = processed.reduce((s, c) => s + c.Gi * c.xi, 0);

  const E = 8.36 * Vt * Gt;
  const K = 10.0 * Vt * Vt * Gt;
  const S = (3 * K) / (10.2 * Vt - 1);
  const L = K + (4 * S) / 3;
  const nu = 0.5 - 1 / (7.2 * Vt);

  return {
    summary: {
      total_coefficient: totalCoeff,
      sumXiVi,
      Vm,
      Vt,
      Gt,
      E,
      K,
      S,
      L,
      nu,
    },
    compounds: processed,
  };
}
